### PR TITLE
DiscordHook improvement and fixing

### PR DESCRIPTION
**Pull request by Tomut0**
The continue of #279 .
We found one more issue and going to fix it. 🚀 

### Fix the clan creation bugs
We decided to leave our stone cave and found some issues with channels creation.
Then we took our cudgel and hit with all our might. Let's believe it works now. 👍🏻 

### Leader role and permissions
Finally, the leader role and view permission is given and taken rightly
(we hope)

### A new config option
```yml
min-linked-players-to-create: 3
```
Now you can set how much linked players are needed to create a new channel.
It would reduce the count of unused channels in discord. 

### API
- `createChannel(Category category, String clanTag)` was removed (this is a life)
- A lot of new exceptions (I eat them on my breakfast)
<hr>
🙂 Thanks @RoinujNosde for his help with this PR.

### CODE REVIEW

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package net.sacredlabyrinth.phaed.simpleclans.hooks.discord.exceptions;
+
+import net.sacredlabyrinth.phaed.simpleclans.SimpleClans;
+import org.jetbrains.annotations.NotNull;
+
+import static net.sacredlabyrinth.phaed.simpleclans.managers.SettingsManager.ConfigField.DEBUG;
+
+public class InvalidChannelException extends Exception {
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java]: extends DiscordHookException

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+public class DiscordHookException extends Exception {
+    public DiscordHookException() {
+        super();
+    }
+
+    public DiscordHookException(String message, @NotNull Object... args) {
+        super(String.format(message, args));
+        if (SimpleClans.getInstance().getSettingsManager().is(DEBUG)) {
+            SimpleClans.getInstance().getLogger().warning(String.format(message, args));
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java]: It's unusual for an exception warning to be printed in the constructor. The try/catch is a better place, IMO.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    public InvalidChannelException(String message) {
+        super(message);
+    }
+
+    public InvalidChannelException(String message, @NotNull Object... args) {
+        super(String.format(message, args));
+        if (SimpleClans.getInstance().getSettingsManager().is(DEBUG)) {
+            SimpleClans.getInstance().getLogger().warning(String.format(message, args));
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java]: Same comment as in the other exception

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java [hunk 4]
```diff
@@ -297,18 +326,54 @@ public Category createCategory() {
     }
 
     /**
-     * Creates a new {@link ClanPlayer.Channel} in available categories,
+     * Creates a new {@link ClanPlayer.Channel} in available SimpleClans' categories,
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java]: ClanPlayer.Channel?

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java [hunk 5]
```diff
@@ -471,18 +502,22 @@ private void resetPermissions() {
...
-                forEach(clanPlayer -> updatePermissions(clanPlayer, clanPlayer.getClan(), ADD));
+                forEach(clanPlayer -> updateViewPermission(clanPlayer, ADD));
     }
 
     private void createChannels() {
-        clanTags.removeAll(discordClanTags);
         for (String clan : clanTags) {
             try {
                 createChannel(clan);
             } catch (CategoriesLimitException | ChannelsLimitException ex) {
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java:524]: If the limit is reached, the loop will break here and not process more of the invalid channel logic.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java [hunk 6]
```diff
@@ -512,9 +543,18 @@ private void updateRole(ClanPlayer clanPlayer, DiscordAction action) {
...
         Member member = getMember(clanPlayer);
-        if (clan == null || member == null) {
+        if (clan == null) {
+            return;
+        }
+
+        updateViewPermission(member, clan, action);
+    }
+
+    private void updateViewPermission(@Nullable Member member, @NotNull Clan clan, DiscordAction action) {
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java]: Member could be NotNull and the overloaded method removed.

**Review by RoinujNosde — CHANGES_REQUESTED** [review]

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java (hunk 6 above)
Comment (Tomut0) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java]: :concern:

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java (hunk 1 above)
Comment (Tomut0) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java]: If I'll extend it, it would break the logic [here](https://github.com/RoinujNosde/SimpleClans/pull/281/files#diff-59dd7bda205c34ef18d1a41d5558eadb390d259e9ad3ea256dfc8e58ad6881ecR582)

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java (hunk 2 above)
Comment (Tomut0) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java]: > It's unusual for an exception warning to be printed in the constructor.

I can agree with it, but isn't it free our code base? 🤔

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java (hunk 1 above)
Comment (RoinujNosde) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/InvalidChannelException.java]: How?

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java (hunk 2 above)
Comment (RoinujNosde) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java]: Can you rephrase that question?

Comment (Tomut0) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java]: With the constructor the code will be shorter. 👀

Comment (RoinujNosde) [reply src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/exceptions/DiscordHookException.java]: Not really.
SimpleClans has a debug (in the main class) method, if used in a try/catch it would be even shorter.

File: src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java [hunk 7]
```diff
@@ -319,7 +396,17 @@ public void createChannel(@NotNull String clanTag)
             throw new CategoriesLimitException();
         }
 
-        createChannel(availableCategory, clanTag);
+        TextChannel textChannel = availableCategory.createTextChannel(clanTag).complete();
+
+        for (Map.Entry<ClanPlayer, Member> entry : discordClanPlayers.entrySet()) {
+            Clan entryClan = entry.getKey().getClan();
+            if (entryClan == null) {
```
Comment (RoinujNosde) [inline src/main/java/net/sacredlabyrinth/phaed/simpleclans/hooks/discord/DiscordHook.java]: Possibly a unnecessary check, because the map is formed from clan#getMembers (so the clan exists).
(The IDE warning can be ignored)

**Review by RoinujNosde — APPROVED** [review]
